snowflake/snowflake.go

```diff
@@ -0,0 +1,256 @@
+func (g *Generator) setEpoch(epoch time.Time) (err error) {
+	const op errors.Op = "snowflakeGenerator.setEpoch"
+	epoch = epoch.UTC()
+
+	if epoch.IsZero() {
+		err = errors.New(op, "Epoch cannot be a zero value")
+		return
+	}
+
+	if epoch.After(time.Now()) {
+		err = errors.New(op, "Epoch cannot be in the future")
+		return
+	}
+
+	// Because s must after now, so the `df` not < 0.
+	if since(epoch) > maxTimestamp {
+		err = errors.New(op, "The maximum life cycle of the snowflake algorithm is 179 years(2^39-10ms)")
+		return
+	}
+	g.epochGuard.Do(func() {
+		g.Epoch = epoch
+	})
+	return
+}
+
+// setResolver set the sequence custom Resolver
+func (g *Generator) setResolver(sequenceResolver SequenceResolver) (err error) {
+	const op errors.Op = "snowflakeGenerator.setResolver"
+	if sequenceResolver == nil {
+		err = errors.New(op, "SequenceResolver cannot be nil")
+		return
+	}
+	g.resolverGuard.Do(func() {
+		g.SequenceResolver = sequenceResolver
+	})
+	return
+}
+
+// setNodeIDProvider set the sequence NodeID provider
+func (g *Generator) setNodeIDProvider(nodeIDProvider NodeIDProvider) (err error) {
+	const op errors.Op = "snowflakeGenerator.setNodeIDProvider"
+	if nodeIDProvider == nil {
+		err = errors.New(op, "NodeIDProvider cannot be nil")
+		return
+	}
+	g.nodeIDProviderGuard.Do(func() {
+		g.NodeIDProvider = nodeIDProvider
+		var nodeID uint16
+		nodeID, err = g.NodeIDProvider()
+		if err != nil {
+			err = errors.New(op, "error generating nodeID", err)
+			return
+		}
+		g.nodeID = nodeID
+	})
+	return
+}
+
+// Decompose returns a map of snowflake id parts.
+func (g *Generator) Decompose(sid uint64) (id ID) {
+	id.MSB = int8(sid >> 63)
+	timestamp := sid >> timeShift
+	id.Timestamp = time.Unix(0, (g.Epoch.UnixNano())+int64(timestamp*timeUnit))
+	id.Sequence = uint16(sid & maskSequence)
+	id.NodeID = uint16(sid & maskNodeID >> nodeShift)
+	return
+}
+
+// NextID generate a new unique snowflake ID
+func (g *Generator) NextID() (sid uint64, err error) {
+	var current int64
+	var seq uint16
+	current, err = g.currentTimestamp()
+
+	if err != nil {
+		return
+	}
+
+	seq, err = g.SequenceResolver(current)
+
+	if err != nil {
+		return
+	}
+
+	for seq >= maxSequence {
+		current, err = g.waitForNext10Millis(current)
+		if err != nil {
+			return
+		}
+
+		seq, err = g.SequenceResolver(current)
+		if err != nil {
+			return
+		}
+	}
+	sid = uint64(current)<<timeShift | uint64(g.nodeID)<<nodeShift | uint64(seq)
+	return
+}
+
+func (g *Generator) currentTimestamp() (current int64, err error) {
+	const op errors.Op = "snowflake_id.currentTimestamp"
+	current = g.currentTimeSlot()
+	if current < 0 || current > maxTimestamp {
+		err = errors.New(op, "timestamp exceeds max time(2^39-1 * 10ms), please check the epoch settings")
+		return
+	}
+	return
+}
+
+func (g *Generator) currentTimeSlot() int64 {
+	return time.Since(g.Epoch).Nanoseconds() / timeUnit
+}
+
+func since(t time.Time) int64 {
+	return time.Since(t).Nanoseconds() / timeUnit
```

Comment:
Is there a special reason not to use milliseconds instead of nanoseconds?
E.g. 
```
time.Since(t).Miliseconds() / timeUnit
```
with `timeUnit = 10*time.Milisecond`

Reply:
Actually no particular reason, can use `Miliseconds()`, updated.